src/core/PdbParser.cpp

```diff
@@ -79,6 +79,8 @@ bool PdbParser::parse_pdb_file(const string &filename) {
 bool PdbParser::parse_itp_file(const string &filename) {
   ifstream file(filename.c_str());
   string tmp, buf;
+  const decltype(buf)::value_type EOF_value =
```

Comment:
Is there is specific reason for using `decltype` instead of just using `string` which is hard-coded one line above? Also you could use auto so you don't have to repeat the type.

Reply:
I tried `const auto EOF_value` first, but the auto is deduced as a signed int.  Not sure why. `decltype(buf)::value_type` is a char that cannot be cast to a signed int on the previously mentioned architectures. Here is a minimum working example:
```c++
#include <iostream>
#include <string>
int main(void) {
  std::string buf = "abcd";
  std::cout << typeid(unsigned char).name() << ": unsigned char" << std::endl;
  std::cout << typeid(signed char).name() << ": signed char" << std::endl;
  std::cout << typeid(char).name() << ": char" << std::endl;
  std::cout << typeid(unsigned int).name() << ": unsigned int" << std::endl;
  std::cout << typeid(signed int).name() << ": signed int" << std::endl;
  std::cout << typeid(int).name() << ": int" << std::endl;
  const auto EOF_value1 =
      std::char_traits<decltype(buf)::value_type>::eof();
  const decltype(buf)::value_type EOF_value2 =
      std::char_traits<decltype(buf)::value_type>::eof();
  std::cout << typeid(decltype(EOF_value1)).name() << ": const auto" << std::endl;
  std::cout << typeid(decltype(EOF_value2)).name() << ": const decltype(buf)::value_type" << std::endl;
  return 0;
}
```
Output:
```
h: unsigned char
a: signed char
c: char
j: unsigned int
i: signed int
i: int
i: const auto
c: const decltype(buf)::value_type
```

Reply:
EOF is not of `value_type`, that is kind of the point. Also you only ever need to compare it to `value_type`,
I don't really see the problem. Also `std::char_traits<char>::int_type` (the return type of eof) is standardized as `int`, so chance are that the eof value will not fit into a char. (see https://en.cppreference.com/w/cpp/string/char_traits).

Reply:
If think the proper way to do it is to use `std::char_traits::not_eof` anyway, instead of an `int` comparison.

Reply:
`buf[0]` is definitely of type `char`, so the comparison can only ever be true if `eof` fits into a `char`.

Reply:
Yes, but it shouln'd be. `std::ifstream::get` returns an `int_type`, which is either a valid character, or `eof`, also an  `int_type`. This result can then be checked for `eof`.

Reply:
I think we can all agree that this API is horrible, but if we want to make it portable, we can't make implicit assumptions such as that the result of `get` or `eof` fit into a char.

Reply:
The issue here is not about the actual value `EOF` since it is -1 by definition, and -1 == 255 (mod 256). How the compiler interprets the signedness of a char is also irrelevant, because the comparison operator will cast the value correctly. Without the -Wextra flag, it compiles just fine and the PdbParser unit test passes. I think the warning just says the compiler might try to optimize away part of the conditional based on the range of values it can hold before the cast. The manual cast of `EOF` helps the compiler figuring out what the developer's intention is.

Reply:
Hmm... let me take that back: the conditional doesn't cast correctly, `EOF` is not detected at all. Let's use `ifstream::eof()` instead.